src/platform_linux/platform_inline.h

```diff
@@ -150,22 +150,32 @@ platform_semaphore_try_wait(platform_semaphore *sema)
 }
 
 static inline platform_status
-platform_mutex_lock(platform_mutex *mu)
-{
-   int ret;
-
-   ret = pthread_mutex_lock(mu);
+platform_mutex_lock(platform_mutex *lock)
+{
+   int ret = pthread_mutex_lock(&lock->mutex);
+   platform_assert(
+      lock->owner == INVALID_TID,
+      "Locked a mutex with an existing owner, lock: %p, tid: %lu, owner: %lu\n",
+      lock,
+      platform_get_tid(),
+      lock->owner);
+   lock->owner = platform_get_tid();
+   platform_assert(lock->owner == platform_get_tid());
 
    return CONST_STATUS(ret);
 }
 
 static inline platform_status
-platform_mutex_unlock(platform_mutex *mu)
-{
-   int ret;
-
-   ret = pthread_mutex_unlock(mu);
-
+platform_mutex_unlock(platform_mutex *lock)
+{
+   platform_assert(lock->owner == platform_get_tid(),
+                   "Attempt to unlock a mutex without ownership, lock: %p, "
+                   "tid: %lu, owner: %lu\n",
+                   lock,
+                   platform_get_tid(),
+                   lock->owner);
+   lock->owner = INVALID_TID;
```

Comment:
I've seen, in past implementations, that calls to these kinds of lock / unlock APIs proliferate, and then we run into lock-ownership sanity checks / doubts.

I see the `platform_assert()` on L171 as a way of our own book-keeping to make sure that lock / unlock sequences are kosher.

Here's another way to cross-check this more deeply. In the `pthread_mutex_lock()` [ref-man](https://pubs.opengroup.org/onlinepubs/007908799/xsh/pthread_mutex_lock.html), I see this mentioned:

```
The pthread_mutex_lock() function may fail if: 

[EDEADLK]
    The current thread already owns the mutex.
```

What if we add a `debug_assert(pthread_mutex_lock(&lock->mutex) == EDEADLK)` ?

Would n't that be a further validation that, on top of our own internal book-keeping, the pthread-lock structures themselves re-confirm that we **are truly** owning the lock?

Reply:
So in contravention to the man page, this just deadlocks on my system. I'm going to get rid of it.